sonic-net/sonic-utilities
# [dpu_tty]: Add a DPU TTY console utility

* Add dpu_tty.py
* Add dpu_tty.py to scripts

Signed-off-by: Wenchung Wang vincwang@cisco.com



#### What I did

Add DPU TTY console utility.

#### How I did it

* Read platform.json to retrieve TTY settings in DPUS section. Set up TTY console according to TTY device name and baud rate.
* Also provide options to overwrite default TTY device and baud rate.

```
"DPUS": {
    "dpu0": {
        "serial-console": {
            "device": "ttyS4",
            "baud-rate": "115200"
        }
    },
    "dpu1": {
        "serial-console": {
            "device": "ttyS5",
            "baud-rate": "115200"
        }
    }
},
```

#### How to verify it

Run the utility on a smart switch that provides DPU UART console via ttyS device. The test plan is at
https://github.com/sonic-net/sonic-mgmt/pull/12701/files section 1.4.

The utility works only on smart switch that provides DPU serial console connections through TTY /dev/ttyS* devices, for example Mt Fuji. Here is a test log.

```
root@MtFuji:/home/cisco# dpu-tty.py -n dpu0
picocom v3.1

port is        : /dev/ttyS4
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        : 
omap is        : 
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready

sonic login: admin
Password: 
Linux sonic 6.1.0-11-2-arm64 #1 SMP Debian 6.1.38-4 (2023-08-08) aarch64
You are on
  ____   ___  _   _ _  ____
 / ___| / _ \| \ | (_)/ ___|
 \___ \| | | |  \| | | |
  ___) | |_| | |\  | | |___
 |____/ \___/|_| \_|_|\____|

-- Software for Open Networking in the Cloud --

Unauthorized access and/or use are prohibited.
All access and/or use are subject to monitoring.

Help:    https://sonic-net.github.io/SONiC/

Last login: Thu Sep 19 18:40:28 UTC 2024 on ttyS0
admin@sonic:~$ show ip int
Interface    Master    IPv4 address/mask    Admin/Oper    BGP Neighbor    Neighbor IP
-----------  --------  -------------------  ------------  --------------  -------------
Ethernet0              18.0.202.1/31        up/up         N/A             N/A
docker0                240.127.1.1/24       up/down       N/A             N/A
int_mnic0              169.254.200.0/24     up/up         N/A             N/A
lo                     127.0.0.1/16         up/up         N/A             N/A
admin@sonic:~$ 
Terminating...
Thanks for using picocom
root@MtFuji:/home/cisco# dpu-tty.py -n dpu7
picocom v3.1

port is        : /dev/ttyS11
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        : 
omap is        : 
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready

sonic login: 
Terminating...
Thanks for using picocom
root@MtFuji:/home/cisco# dpu-tty.py -n dpu7
picocom v3.1

port is        : /dev/ttyS11
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        : 
omap is        : 
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready

sonic login: admin
Password: 
Linux sonic 6.1.0-11-2-arm64 #1 SMP Debian 6.1.38-4 (2023-08-08) aarch64
You are on
  ____   ___  _   _ _  ____
 / ___| / _ \| \ | (_)/ ___|
 \___ \| | | |  \| | | |
  ___) | |_| | |\  | | |___
 |____/ \___/|_| \_|_|\____|

-- Software for Open Networking in the Cloud --

Unauthorized access and/or use are prohibited.
All access and/or use are subject to monitoring.

Help:    https://sonic-net.github.io/SONiC/

admin@sonic:~$ sudo su
root@sonic:/home/admin# show ip int
Interface    Master    IPv4 address/mask    Admin/Oper    BGP Neighbor    Neighbor IP
-----------  --------  -------------------  ------------  --------------  -------------
Ethernet0              18.7.202.1/31        up/up         N/A             N/A
docker0                240.127.1.1/24       up/down       N/A             N/A
int_mnic0              169.254.200.7/24     up/up         N/A             N/A
lo                     127.0.0.1/16         up/up         N/A             N/A
root@sonic:/home/admin# 
Terminating...
Thanks for using picocom
root@MtFuji:/home/cisco# 
```

#### Previous command output (if the output of a command-line utility has changed)

N/A

#### New command output (if the output of a command-line utility has changed)

N/A
